libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -117,11 +116,44 @@ proc handle*(p: PubSubPeer, conn: Connection) {.async.} =
       debug "exiting pubsub peer read loop"
       await conn.close()
 
+      if p.sendConn == conn:
+        p.sendConn = nil
+
   except CancelledError as exc:
     raise exc
   except CatchableError as exc:
     trace "Exception occurred in PubSubPeer.handle", exc = exc.msg
 
+proc getSendConn(p: PubSubPeer): Future[Connection] {.async.} =
+  # get a cached send connection or create a new one - need to call
+  # `putSendConn` after use! If one is not available, set up a new one
+  var res = p.sendConn
+
+  if not res.isNil:
+    if res.closed() or res.atEof:
+      p.sendConn = nil
+      await res.close() # p.sendConn might change after here
+      res = nil
+
+  if res.isNil:
+    # get a send connection if there is none
+    res = await p.switch.dial(p.peerId, p.codec) # ...and here
+    if res == nil:
+      return nil
+
+    # It so happens that there may be multiple dial attempts in-flight here - we
+    # also cannot take a lock because that might block mplex backpressure
+    # handling - thus we might redundantly create more send connections than
+    # would normally be needed - we'll cycle through them as we move along..
+    p.sendConn = res
```

Comment:
I don't quite understand the reasoning here. Why don't you handle the case where `p.sendConn` is not nil here? (it was set by a concurrent `dial` attempt). It would be easy to close the redundant connection, return the established one and then move on.

Reply:
it's more or less the same: we have two send connections thus one of them has to be removed - doesn't really matter which?

Reply:
alright, I changed it to reuse the existing connection - it's actually slightly more simple that way because the existing connection cannot be closed at once as it might still be used in another concurrent send call..